### PR TITLE
Transactions management for RxJava1 and RxJava2

### LINKED ISSUE #141: Transaction management for SQLClient
When using an Rxified SQLClient, it's not obvious for everyone how to properly manage a connection. Proper transaction management is even worse.

It would help to have utility methods to manage a connection, a transaction, or both.

Let's take an example:

```java
public class JDBCHelper {

  static <T> Flowable<T> withConnectionFlowable(SQLClient sqlClient, Function<SQLConnection, Flowable<T>> flowable) {
    return sqlClient.rxGetConnection().flatMapPublisher(sqlConnection -> {
      return flowable.apply(sqlConnection).doFinally(sqlConnection::close);
    });
  }

  static <T> Flowable<T> inTransactionFlowable(SQLClient sqlClient, Function<SQLConnection, Flowable<T>> flowable) {
    return withConnectionFlowable(sqlClient, flowable);
  }

  static <T> Flowable<T> inTransactionFlowable(SQLConnection sqlConnection, Function<SQLConnection, Flowable<T>> flowable) {
    return sqlConnection.rxSetAutoCommit(false).<T>toFlowable()
      .concatWith(flowable.apply(sqlConnection))
      .concatWith(sqlConnection.rxCommit().toFlowable())
      .onErrorResumeNext(throwable -> {
        return sqlConnection.rxRollback().onErrorComplete().andThen(Flowable.error(throwable));
      }).concatWith(sqlConnection.rxSetAutoCommit(true).toFlowable());
  }
}
```

Now it's easy to write such kind of flow:

```java
    JDBCHelper.withConnectionFlowable(client, conn -> {
      Flowable<JsonArray> setupDb = JDBCHelper.inTransactionFlowable(conn, sqlConnection -> {
        return conn.rxUpdateWithParams("INSERT INTO colors (name) VALUES (?)", new JsonArray().add("BLACK")).toCompletable().toFlowable()
          .concatWith(conn.rxUpdateWithParams("INSERT INTO colors (name) VALUES (?)", new JsonArray().add("WHITE")).toCompletable().toFlowable())
          .concatWith(conn.rxUpdateWithParams("INSERT INTO colors (name) VALUES (?)", new JsonArray().add("PURPLE")).toCompletable().toFlowable());
      }).ignoreElements().toFlowable();
      return setupDb
        .concatWith(conn.rxQuery("SELECT * FROM colors").flatMapPublisher(resultSet -> Flowable.fromIterable(resultSet.getResults())));
    }).subscribe(System.out::println, Throwable::printStackTrace);
```

Of course, we'd need other helper methods for `Single`, `Completable` and `Maybe`.

Also, it's not clear to me where we'd add this code. In this repo? In `vertx-sql-common`? Should we use utility classes as in the example above? Or should we add these methods to Rxified `SQLClient` classes? In this case, how?

### PR DESCRIPTION
**Pull request by tsegismont**
Fixes #141 

Managing a transaction with the Rxified API requires quite a lot of [boilerplate](https://github.com/vert-x3/vertx-examples/blob/1ce6051c6de0af846a78e13969ae08446b74dc73/rxjava-1-examples/src/main/java/io/vertx/example/rxjava/database/jdbc/Transaction.java).

This PR provides transformers for all RxJava1 and RxJava2 constructs, so that adding transaction management becomes a matter of invoking the `compose` operator. For example:

```java
    Flowable<JsonArray> flowable = client.rxGetConnection().flatMapPublisher(conn -> {
      return conn.rxExecute(insert1)
        .andThen(conn.rxExecute(insert2))
        .andThen(conn.rxQuery(select).flatMapPublisher(resultSet -> Flowable.fromIterable(resultSet.getResults()))
        .compose(new InTransactionFlowable<>(conn))
        .doFinally(conn::close);
    });
```

### CODE REVIEW

**Comment by tsegismont** [comment]
@vietj can you please take a look?

**Comment by vietj** [comment]
not yet reviewed all, but one thing you need to make sure is that the RxJavaPlugins are managed correctly.

**Comment by tsegismont** [comment]
wdym ?

Le mar. 29 mai 2018 à 11:38, Julien Viet <notifications@github.com> a
écrit :

**Comment by vietj** [comment]
for instance we had this PR contributed that fixes a few that I did not have : https://github.com/vert-x3/vertx-rx/pull/134

in short we want to be sure it can be used for context tracking via Rx plugins.

**Comment by tsegismont** [comment]
@vietj I looked into #134. The issue does not apply here. There is no observable created, just using regular observable api methods (`compose`, `flatMap`, ...etc). These methods take care of invoking plugin hooks.

**Comment by vietj** [comment]
since the user does not need to know the  `InTransactionCompletable`, etc... types, we should move these into `impl` package and make static factory method on an helper class.

**Comment by vietj** [comment]
I can't see any documentation.

**Comment by tsegismont** [comment]
>since the user does not need to know the  InTransactionCompletable, etc... types

Helper methods can be useful for users who need a single transactional flow. For `Flowable` it could look like:

```java
<T> Flowable<T> inTransaction(SQLClient client, Function<SQLConnection, Flowable<T>> sourceSupplier);
```

I will add such helpers for all observable flavors.

However some users might want to manage the connection themselves (eg if they do operations in autocommit mode before or after the transactional flow). For such use cases, using `compose` and the individual transformers provides maximum flexibility.

> I can't see any documentation.

Will do

**Comment by jponge** [comment]
I like it!

**Comment by FroMage** [comment]
Yeah, this sounds like a great feature!

**Comment by tsegismont** [comment]
@vietj done

@jponge can you take a look at the commit at the documentation commit?

File: rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc [hunk 1]
```diff
@@ -0,0 +1,36 @@
+== Transaction management with RxJava
+
+Managing transactions with the Rxified API `SQLConnection` requires quite a lot of boilerplate code.
+
+Vert.x provides observable __transformers__ that you can apply to your flows with `compose` in order to make them transactional:
```
Comment (jponge) [inline rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc]: I'd put "observable transformers" in the same emphasis.

Comment (jponge) [reply rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc]: You may also link to the `compose` operator in RxJava for the interested reader.

File: rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+* `InTransactionObservable`
+* `InTransactionSingle`
+* `InTransactionCompletable`
+
+These _transformers_ wrap the corresponding source of events with transaction management. Let's take an example.
```
Comment (jponge) [inline rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc]: with _SQL_ transaction management.

File: rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+In your music library application, you need to insert a row in table `albums`, then some rows in table `tracks`.
+These two steps shall be part of the same transaction.
```
Comment (jponge) [inline rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc]: _atomic_ transaction.

File: rx-java/src/main/java/examples/SQLExamples.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
+/*
```
Comment (jponge) [inline rx-java/src/main/java/examples/SQLExamples.java]: ❤️ Ah finally someone who cares about license headers 🤣

File: rx-java2/src/main/asciidoc/vertx-rx/java2/sql.adoc [hunk 5]
```diff
@@ -0,0 +1,38 @@
+== Transaction management with RxJava
```
Comment (jponge) [inline rx-java2/src/main/asciidoc/vertx-rx/java2/sql.adoc]: Same remarks as for the Rx1 doc

**Comment by tsegismont** [comment]
@jponge docs fixed

File: rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc [hunk 6]
```diff
@@ -0,0 +1,36 @@
+== SQL Transaction management with RxJava
+
+Managing SQL transactions manually with the Rxified `SQLConnection` requires quite a lot of boilerplate code.
+
+Vert.x provides __observable transformers__ that you can apply to your flows with http://reactivex.io/RxJava/1.x/javadoc/rx/Observable.html#compose-rx.Observable.Transformer-[`compose`] in order to make them transactional:
```
Comment (vietj) [inline rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc]: I think _in order_ is not necessary

File: rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+
+* `InTransactionObservable`
+* `InTransactionSingle`
+* `InTransactionCompletable`
+
+These _transformers_ wrap the corresponding source of events with SQL transaction management. Let's take an example.
+
+In your music library application, you need to insert a row in table `albums`, then some rows in table `tracks`.
+These two steps shall be part of the same atomic transaction.
```
Comment (vietj) [inline rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc]: transactions are always atomic :-)

File: rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+If it succeeds, the application must return results from a query involving both tables.
+
+After your got an instance of `io.vertx.rxjava.ext.sql.SQLConnection`, you may use it to perform the SQL operations:
```
Comment (vietj) [inline rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc]: _you can use it_

File: rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+
+[source,java]
+----
+include::SQLExamples.java[tags=inTransactionTransformer]
```
Comment (vietj) [inline rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc]: we need to use here the inclusion with docgen

**Comment by vietj** [comment]
Hi, I've done a bit of reviews, for the examples we need to use docgen and not asciidoc inclusion, the main reason is that docgen will produce an asciidoc file containing the javacode inserted and that will be used after for producing web-site documentation.

File: rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc (hunk 7 above)
Comment (tsegismont) [reply rx-java/src/main/asciidoc/vertx-rx/java/sql.adoc]: I guess the point was to insist on the A of ACID here

**Comment by tsegismont** [comment]
@vietj done

**Comment by tsegismont** [comment]
ok to merge?

**Comment by vietj** [comment]
I need to make one more pass on the review

**Comment by vietj** [comment]
I think we need to move the various Transformer implementation to an impl package and make static method on the helper to create them. Such transformers don't expose method so there is no point the user need to know about the class implementing. That's what we do for the vertx-core helpers.

**Comment by tsegismont** [comment]
I disagree. See https://github.com/vert-x3/vertx-rx/pull/143#issuecomment-393600598

It's not an implementation detail. It's about flexibility

**Comment by vietj** [comment]
why is it more flexible to do `Transformer transformer = new A()` rather than `Transformer transformer  = Helper.a()` ?

**Comment by tsegismont** [comment]
I misunderstood your comment, I thought you wanted to hide the transformer completely, and only expose helper methods.

So IIUC, you'd like to see something similar to `io.vertx.reactivex.impl.FlowableUnmarshaller` created by `io.vertx.reactivex.core.FlowableHelper#unmarshaller(java.lang.Class<T>)`?

**Comment by vietj** [comment]
Simply:

```
public static Transformer inTransactionCompletable(SQLConnection conn) {
   return new InTransactionCompletable(conn);
}
```

**Comment by tsegismont** [comment]
OK

**Comment by tsegismont** [comment]
@vietj done

**Comment by vietj** [comment]
@tsegismont LGTM! thanks for improving usability of painful transactions!

**Pull request merged**